gtsfm/frontend/matcher/twoway_matcher.py

```diff
@@ -86,3 +85,55 @@ def match(
         match_indices[:, 1] = valid_idx_i2[match_indices[:, 1]]
 
         return match_indices
+
+    def __init_opencv_matcher(self) -> cv.DescriptorMatcher:
+        """Initialize the OpenCV matcher.
+
+        Returns:
+            Matcher object.
+        """
+        if self._distance_type is MatchingDistanceType.EUCLIDEAN:
+            distance_metric = cv.NORM_L2
+        elif self._distance_type is MatchingDistanceType.HAMMING:
+            distance_metric = cv.NORM_HAMMING
+        else:
+            raise NotImplementedError("The distance type is not in MatchingDistanceType")
+
+        return cv.BFMatcher(normType=distance_metric, crossCheck=False)
+
+    def __perform_matching(self, descriptors_1: np.ndarray, descriptors_2: np.ndarray) -> np.ndarray:
+        """Run the core logic for matching.
+
+        Args:
+            descriptors_1: descriptors for the 1st image.
+            descriptors_2: descriptors for the 2nd image.
+
+        Returns:
+            indices of the match between two images.
+        """
+        match_indices_1to2: Dict[int, int] = self.__perform_oneway_matching(descriptors_1, descriptors_2)
+        match_indices_2to1: Dict[int, int] = self.__perform_oneway_matching(descriptors_2, descriptors_1)
+
+        match_indices_1to2to1 = {
+            idx1: match_indices_2to1[idx2] for idx1, idx2 in match_indices_1to2.items() if idx2 in match_indices_2to1
+        }
+
+        twoway_match_indices = np.array(
+            [[idx1, idx2] for idx1, idx2 in match_indices_1to2.items() if match_indices_1to2to1.get(idx1) == idx1],
+            dtype=np.uint32,
+        )
```

Comment:
I think this can be accomplished without looping through all the matches twice.

```python
matches = [(idx1, idx2) for idx1, idx2 in match_ind_1to2.items() if match_ind_2to1[idx2] == idx1]
```